sdk/rms_sdk/PDFObjectModel/PDFCreator.cpp

```diff
@@ -0,0 +1,672 @@
+  if (!root_dictionary) return false;
+
+  acro_form_dictionary = root_dictionary->GetDict("AcroForm");
+  if (!acro_form_dictionary) return false;
+
+  if (!acro_form_dictionary->KeyExist("XFA")) {
+    return false;
+  }
+
+  CPDF_Object* pdf_xfa_object = acro_form_dictionary->GetElement("XFA");
+  CPDF_Array* pdf_xfa_array = pdf_xfa_object->GetArray();
+  if (!pdf_xfa_array) return false;
+
+  FX_DWORD count = pdf_xfa_array->GetCount();
+  for (int i = 0; i < count; i++) {
+    CPDF_Object* pdf_object = pdf_xfa_array->GetElement(i);
+    if (NULL != pdf_object && pdf_object->GetType() == PDFOBJ_REFERENCE) {
+      CPDF_Object* pdf_reference_object = (CPDF_Object*)pdf_object->GetDict();
+      if (NULL != pdf_reference_object) {
+        CPDF_Dictionary* pdf_dictionary = pdf_reference_object->GetDict();
+        if (NULL != pdf_dictionary && pdf_dictionary->GetInteger("Length") != 0) {
+          if (root_dictionary->KeyExist("NeedsRendering")) {
+            bool dynamic_xfa = root_dictionary->GetBoolean("NeedsRendering");
+            if (dynamic_xfa) {
+              return true;
+            } else {
+              //is static XFA
+              return false;
+            }
+          } else {
+            //is static XFA
+            return false;
+          }
+        }
+      }
+    }
+  }
+  return false;
+}
+
+uint32_t PDFCreatorImpl::ParsePDFFile(CPDF_Parser *pdf_parser) {
+  uint32_t result = PDFCREATOR_ERR_SUCCESS;
+  CFX_ByteString file_path_bytestring = file_path_.c_str();
+
+  FX_DWORD parse_result = pdf_parser->StartParse(file_path_bytestring);
+  result = ConvertParsingErrCode(parse_result);
+
+  if(result == PDFCREATOR_ERR_SUCCESS) {
+    //has been protected by password
+    if (IsProtectedByPassword(pdf_parser)) {
+      return PDFCREATOR_ERR_SECURITY;
+    }
+
+    //has been signed?
+    if (IsSigned(pdf_parser)) {
+      return PDFCREATOR_ERR_FORMAT;
+    }
+
+    //is dynamic XFA PDF
+    if (IsDynamicXFA(pdf_parser)) {
+      return PDFCREATOR_ERR_FORMAT;
+    }
+  }
+
+  return result;
+}
+
+uint32_t PDFCreatorImpl::ConvertParsingErrCode(FX_DWORD parse_result) {
+  uint32_t result = PDFCREATOR_ERR_SUCCESS;
+  switch (parse_result) {
+    case PDFPARSE_ERROR_SUCCESS: {
+      result = PDFCREATOR_ERR_SUCCESS;
+      break;
+    }
+    case PDFPARSE_ERROR_FILE: {
+      result = PDFCREATOR_ERR_FILE;
+      break;
+    }
+    case PDFPARSE_ERROR_FORMAT: {
+      result = PDFCREATOR_ERR_FORMAT;
+      break;
+    }
+
+    case PDFPARSE_ERROR_PASSWORD:
+    case PDFPARSE_ERROR_HANDLER:
+    case PDFPARSE_ERROR_CERT: {
+      result = PDFCREATOR_ERR_SECURITY;
+      break;
+    }
+
+    default: {
+      result = PDFCREATOR_ERR_UNKNOWN;
+      break;
+    }
+
+  }
+
+  return result;
+}
+
+CPDF_Dictionary* PDFCreatorImpl::CreateEncryptionDict(
+    const std::string& filter_name,
+    const std::vector<unsigned char> &publishing_license) {
+  CPDF_Dictionary* encryption_dictionary = new CPDF_Dictionary;
+
```

Comment:
use smart pointers where memory is being allocated. For example with 'new' keyword.

Reply:
This interface is inherited from foxit core interface. Foxit pdf core takes over core_pdf_crypto_handler. Using smart pointer here is not good for performance.